docs/experiments/taint-propagators.md

```diff
@@ -49,14 +49,13 @@ pattern-propagators:
 ```
 
 In the example above, whenever Semgrep finds the pattern `$X.add($Y)`, it checks if
-`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. Thus,
-adding tainted data to a set marks the set itself as tainted.
+`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. 
+Thus, adding tainted data to a set marks the set itself as tainted.
 
 <iframe src="https://semgrep.dev/embed/editor?snippet=7lNe" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-## Propagator rule requirements
+## Requirements of taint propagator rules
 
-A propagator rule necessitates:
+A taint propagator rules necessitate:
 - Any arbitrary pattern that must bind two metavariables of your choice.
-- You must specify the `from` and the `to` metavariables.
-
+- The `from` and `to` metavariables. The `from` indicates the starting variable from which the tainted source passes through to propagate the taint. The `to` specifies what additional object or structure is tainted as a result.
```

Comment:
Actually neither `from` nor `to` are restricted to match variables, and `from` in particular may not be a variable. The `to` will more often be a variable.

Reply:
@IagoAbal so what does the `from` specify mostly? Feel free to suggest your wording of the sentence, as it will reduce the guess work.

Reply:
I mean, this answer needs more specification. @IagoAbal

Reply:
@IagoAbal why didn't you say this when reviewing my video script??

If this is the case, what are situations where people are using it not involving variables? Especially `from`? Why would someone use it that way, and how would you describe it?

As @adamkvitek points out, not adding any information other than saying "that's incorrect" increases the amount of back and forth on these sorts of reviews, which was brought up in the meeting yesterday as being stressful.

Reply:
Initially it was restricted to variables, then it was generalized. In general `from` and `to` can match any expression, but being variables should be the most common case, specially for `to`. The feature is new and experimental, so we don't know how people will be using it, but I saw the use case of tainting a data structure several times, hence I added the feature. The video is not incorrect!

Reply:
@efortuna @IagoAbal thank you! I am suggesting edits below - please let me know if this is correct.

Should I substitute the word "structure" with "data structure" to be consistent (it is used in the introduction also)?

```suggestion
- The `from` and `to` metavariables.
    - The `from` indicates the starting variable or expression from which the tainted source passes through to propagate the taint. 
    - The `to` specifies what additional object or data structure, most commonly a variable, is tainted as a result.
```

Reply:
@IagoAbal can you provide a suggestion for wording so that someone coming to this for the first time would know how to conceptualize `from` and `to` if they can be anything?

Reply:
The use of the term "taint source" here is a bit confusing to me, because a "taint source" has a special meaning in taint-mode. It's not the "tainted source" that passes through the propagator, but the "taint". The "taint source" is the origin of the taint, e.g. `dangerous_user_input`.

You may want to rewrite it but this would be my proposal:

> The `from` metavariable specifies the _entry point_ of the propagator, it should match a variable, or generally an expression. The `to` metavariable specifies the _exit point_, it should also match a variable (or generally an expression), and typically denotes an object or data structure. Taint that reaches the entry point (the `from`) will be propagated to the exit point (the `to`) by the taint engine.

Reply:
Thanks, @IagoAbal, for this great information.

By " it should match a variable" do you mean the `from` metavariable:

A. Matches a variable or expression.
B. A user should specify it as a variable or expression.
C. Both A. and B. (I guess you mean this one.) When a user specifies it as a variable or expression - Semgrep matches variables or expressions with it. I will use the imperative word "specify" below, so the reader is pushed to action here. Let me know if it is correct.

Working with your suggestion:

```suggestion
- The `from` and `to` metavariables.
    - The `from` metavariable specifies the entry point of the taint propagator. Specify a value of `from` metavariable as a variable, or generally as an expression. 
    - The `to` metavariable specifies the exit point of the taint propagator. Specify a value of `to` metavariable as a variable, or generally as an expression. This metavariable typically denotes an object or data structure. Taint that reaches the entry point (the `from`)  is propagated to the exit point (the `to`) by the taint engine.
```

Reply:
I don't think we need the an additionally indented list. 

```suggestion
- A `from` metavariable specifying the entry point of the taint. 
- A `to` metavariable specifying where the tainted data is propagated to, typically an object or data structure.

`to` and `from` can contain either a variable or expression.
```

Side note, if we're being pedantic: @IagoAbal You say "I saw the use case of tainting a data structure several times" but a data structure itself is not an expression. You saw people taint a data structure _in_ a `from` or `to` value?

Reply:
```suggestion
- Use `to` and `from` metavariables. Both of these metavariables must contain either a **variable** or an **expression**.
    - The `from` metavariable specifies the entry point of the taint.
    - The `to` metavariable specifies where the tainted data is propagated to, typically an object or data structure.
```